### PR TITLE
Cleanup codegen traits

### PR DESCRIPTION
**Pull request by nnethercote**
The traits governing codegen are quite complicated and hard to follow. This PR cleans them up a bit.

r? @bjorn3

### CODE REVIEW

**Comment by nnethercote** [comment]
Some opinionated changes here, let's see how it goes. Best reviewed one commit at a time.

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -634,7 +634,7 @@ pub type TyAndLayout<'tcx> = rustc_target::abi::TyAndLayout<'tcx, Ty<'tcx>>;
 pub trait LayoutOfHelpers<'tcx>: HasDataLayout + HasTyCtxt<'tcx> + HasParamEnv<'tcx> {
     /// The `TyAndLayout`-wrapping type (or `TyAndLayout` itself), which will be
     /// returned from `layout_of` (see also `handle_layout_err`).
-    type LayoutOfResult: MaybeResult<TyAndLayout<'tcx>>;
+    type LayoutOfResult: MaybeResult<TyAndLayout<'tcx>> = TyAndLayout<'tcx>;
```
Comment (bjorn3) [inline compiler/rustc_middle/src/ty/layout.rs:637]: Is there any case where this is not `TyAndLayout<'tcx>` and `FnAbiResult` is not `&'tcx FnAbi<'tcx, Ty<'tcx>>`?

**Comment by bjorn3** [comment]
I'm not entirely happy with "[Rename supertraits of CodegenMethods.](https://github.com/rust-lang/rust/pull/130457/commits/a8d22eb39eafcc19d4c3db4f24582d869c8fdef2)" Other than that everything looks like an improvement to me.

File: compiler/rustc_middle/src/ty/layout.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_middle/src/ty/layout.rs:637]: For both cases there are several places where it's `Result<TyAndLayout<'tcx>, _>` and `Result<&'tcx FnAbi<'tcx, Ty<'tcx>>, _>`.

**Comment by nnethercote** [comment]
> I'm not entirely happy with "[Rename supertraits of CodegenMethods.](https://github.com/rust-lang/rust/pull/130457/commits/a8d22eb39eafcc19d4c3db4f24582d869c8fdef2)" Other than that everything looks like an improvement to me.

I can remove it, though I would be interested to hear why you don't like it. The `Builder`/`Codegen` symmetry seemed obvious to me...
